### PR TITLE
usb: hid: idle rate

### PR DESCRIPTION
**Pull request by masz-nordic**
Add idle rate functionality in USB HID class.
Includes a rework of handling class-specific requests in HID.

Tested with USB3CV and a host using idle rate.

### CODE REVIEW

File: subsys/usb/class/hid/core.c [hunk 1]
```diff
@@ -89,6 +90,62 @@ static struct hid_device_info {
 	const struct hid_ops *ops;
 } hid_device;
 
+int hid_on_get_idle(struct usb_setup_packet *setup, s32_t *len, u8_t **data)
```
Comment (finikorg) [inline subsys/usb/class/hid/core.c]: Please use static

File: subsys/usb/class/hid/core.c [hunk 2]
```diff
@@ -337,3 +425,4 @@ int hid_int_ep_read(u8_t *data, u32_t max_data_len, u32_t *ret_bytes)
 	return -ENOTSUP;
 #endif
 }
+
```
Comment (finikorg) [inline subsys/usb/class/hid/core.c]: What is this?

File: subsys/usb/class/hid/core.c [hunk 3]
```diff
@@ -86,17 +88,24 @@ USBD_CLASS_DESCR_DEFINE(primary) struct usb_hid_config hid_cfg = {
...
 int hid_on_get_idle(struct usb_setup_packet *setup, s32_t *len, u8_t **data)
 {
-	LOG_DBG("Get Idle callback");
+	u8_t report_id = setup->wValue & 0xFF;
 
-	/* TODO: Do something. */
+	LOG_DBG("Get Idle callback, report_id: %d", report_id);
 
-	return -ENOTSUP;
+	u32_t siz = sizeof(hid_device.idle_rate[report_id]);
```
Comment (finikorg) [inline subsys/usb/class/hid/core.c]: Please use size

File: subsys/usb/class/hid/core.c [hunk 4]
```diff
@@ -120,9 +129,44 @@ int hid_on_get_protocol(struct usb_setup_packet *setup, s32_t *len,
 
 int hid_on_set_idle(struct usb_setup_packet *setup, s32_t *len, u8_t **data)
 {
-	LOG_DBG("Set Idle callback");
+	u8_t rate = ((setup->wValue & 0xFF00) >> 8);
+	u8_t report_id = setup->wValue & 0xFF;
```
Comment (finikorg) [inline subsys/usb/class/hid/core.c]: Do we need to use le16_to_cpu() ?

**Review by finikorg — CHANGES_REQUESTED** [review]
Since not all drivers have support for `USB_DC_SOF` you need to make sure that it is disabled there.

**Review by jfischer-no — CHANGES_REQUESTED** [review]

File: subsys/usb/class/hid/core.c (hunk 1 above)
[the remark this answers is not in the document]
Comment (masz-nordic) [reply subsys/usb/class/hid/core.c]: Those callbacks are dummy for now, but later I will be implementing them (`hid_on_get_idle` and `hid_on_set_idle` are implemented in the next commit).

>  It is also logged twice, in the specific case and in the dummy callback.

I will remove redundant messages from `hid_class_handle_req()`.

File: samples/subsys/usb/hid/src/main.c [hunk 5]
```diff
@@ -56,34 +57,6 @@ static const u8_t hid_report_desc[] = {
 	HID_MI_COLLECTION_END,
 };
 
-static int debug_cb(struct usb_setup_packet *setup, s32_t *len,
```
[the remark this answers is not in the document]
Comment (masz-nordic) [reply samples/subsys/usb/hid/src/main.c]: >Can the change in this file not be done in separate commit?

Of course I can combine both commits, but for me changing the handlers and implementing idle rate seemed distinct enough to separate them.

**Review by pawelzadrozniak — APPROVED** [review]

**Review by finikorg — APPROVED** [review]
LGTM

File: subsys/usb/class/hid/core.c [hunk 6]
```diff
@@ -123,11 +143,54 @@ static int hid_on_get_protocol(struct usb_setup_packet *setup, s32_t *len,
...
+	LOG_DBG("Set Idle callback, rate: %d, report_id: %d", rate, report_id);
+
+	hid_device.idle_rate[report_id] = rate;
 
+	if (rate == 0) {
+		/* Clear idle */
+		bool clear = true;
+		u8_t i = 0;
+
+		for (i = 1; i <= CONFIG_USB_HID_REPORTS; i++) {
```
Comment (jfischer-phytec-iot) [inline subsys/usb/class/hid/core.c]: I think you can declare loop variable here (or is there a new rule against it?). The same on other places below.

File: subsys/usb/class/hid/core.c [hunk 7]
```diff
@@ -123,11 +143,54 @@ static int hid_on_get_protocol(struct usb_setup_packet *setup, s32_t *len,
...
+	}
+
+	LOG_DBG("Set Idle callback, rate: %d, report_id: %d", rate, report_id);
+
+	hid_device.idle_rate[report_id] = rate;
 
+	if (rate == 0) {
+		/* Clear idle */
+		bool clear = true;
+		u8_t i = 0;
```
Comment (jfischer-phytec-iot) [inline subsys/usb/class/hid/core.c]: u8_t or u16_t ? [Below](https://github.com/zephyrproject-rtos/zephyr/pull/11802/commits/751df9599689aa7b20990582f308e8b1b03012f2#diff-ca310b6eb6831a8122b8134d5fdb6ae6R225) it is u16_t

File: subsys/usb/class/hid/core.c (hunk 6 above)
Comment (masz-nordic) [reply subsys/usb/class/hid/core.c]: Done.

File: subsys/usb/class/hid/core.c (hunk 7 above)
Comment (masz-nordic) [reply subsys/usb/class/hid/core.c]: u16_t. Fixed.

**Review by jfischer-no — APPROVED** [review]

**Pull request merged**